core/trino-main/src/main/java/io/trino/execution/buffer/PagesSerdeUtil.java

```diff
@@ -93,60 +100,82 @@ public static long writePages(PagesSerde serde, SliceOutput sliceOutput, Iterato
         return size;
     }
 
-    public static Iterator<Page> readPages(PagesSerde serde, SliceInput sliceInput)
+    public static Iterator<Page> readPages(PagesSerde serde, InputStream inputStream)
     {
-        return new PageReader(serde, sliceInput);
+        return new PageReader(serde, inputStream);
     }
 
     private static class PageReader
             extends AbstractIterator<Page>
     {
         private final PagesSerde serde;
         private final PagesSerde.PagesSerdeContext context;
-        private final SliceInput input;
+        private final InputStream inputStream;
+        private final byte[] headerBuffer = new byte[SERIALIZED_PAGE_HEADER_SIZE];
+        private final Slice headerSlice = Slices.wrappedBuffer(headerBuffer);
 
-        PageReader(PagesSerde serde, SliceInput input)
+        PageReader(PagesSerde serde, InputStream inputStream)
         {
             this.serde = requireNonNull(serde, "serde is null");
-            this.input = requireNonNull(input, "input is null");
+            this.inputStream = requireNonNull(inputStream, "inputStream is null");
             this.context = serde.newContext();
         }
 
         @Override
         protected Page computeNext()
         {
-            if (!input.isReadable()) {
-                context.close(); // Release context buffers
-                return endOfData();
+            try {
+                int read = ByteStreams.read(inputStream, headerBuffer, 0, headerBuffer.length);
```

Comment:
Could this potentially generate a lot of OS sys calls when pages are small?
We now have a read for page size + a read for page data. If pages are tiny (e.g. because there is a lot of nodes) this increase number of OS IO calls a lot.

Reply:
Discussed offline. Netty inputstreams have internal buffering.

Reply:
> Discussed offline. Netty inputstreams have internal buffering.

I don't think it's true though for unspilled data. However, page size should still be bigger than `SliceInput` buffer